include/swift/Sema/CSFix.h

```diff
@@ -3735,14 +3740,43 @@ class AllowConcreteTypeSpecialization final : public ConstraintFix {
     return diagnose(*commonFixes.front().first);
   }
 
-  static AllowConcreteTypeSpecialization *
-  create(ConstraintSystem &cs, Type concreteTy, ConstraintLocator *locator);
+  static AllowConcreteTypeSpecialization *create(ConstraintSystem &cs,
+                                                 Type concreteTy,
```

Comment:
How about we add a `ValueDecl *` here again and diagnose macros based on MacroDeclaration by using its name if it's an external macro? Something like `cannot specialize a non-generic external macro '<name>'`?